ui/src/data_explorer/components/Visualization.js

```diff
@@ -40,13 +47,71 @@ const Visualization = React.createClass({
   },
 
   getInitialState() {
+    const {queryConfigs, activeQueryIndex} = this.props
+    if (!queryConfigs.length || activeQueryIndex === null) {
+      return {
+        view: GRAPH,
+      }
+    }
+
     return {
-      isGraphInView: true,
+      view: typeof queryConfigs[activeQueryIndex].rawText === 'string' ? TABLE : GRAPH,
     }
   },
 
-  handleToggleView() {
-    this.setState({isGraphInView: !this.state.isGraphInView})
+  componentWillReceiveProps(nextProps) {
+    const {queryConfigs, activeQueryIndex} = nextProps
+    if (!queryConfigs.length || activeQueryIndex === null || activeQueryIndex === this.props.activeQueryIndex) {
+      return
+    }
+
+    const activeQuery = queryConfigs[activeQueryIndex]
+    if (activeQuery && typeof activeQuery.rawText === 'string') {
+      return this.setState({view: TABLE})
+    }
+  },
+
+  handleToggleView(view) {
+    this.setState({view})
+  },
+
+  render() {
+    const {queryConfigs, timeRange, height, heightPixels, onEditRawStatus, activeQueryIndex} = this.props
+    const {source} = this.context
+    const proxyLink = source.links.proxy
+    const {view} = this.state
+
+    const statements = queryConfigs.map((query) => {
+      const text = query.rawText || buildInfluxQLQuery(timeRange, query)
+      return {text, id: query.id}
+    })
+    const queries = statements.filter((s) => s.text !== null).map((s) => {
+      return {host: [proxyLink], text: s.text, id: s.id}
+    })
+
+    return (
+      <div className="graph" style={{height}}>
+        <VisHeader views={VIEWS} view={view} onToggleView={this.handleToggleView} name={name || 'Graph'}/>
+        <div className={classNames({"graph-container": view === GRAPH, "table-container": view === TABLE})}>
+          {this.renderVisualization(view, queries, heightPixels, onEditRawStatus, activeQueryIndex)}
+        </div>
+      </div>
+    )
+  },
+
+  renderVisualization(view, queries, heightPixels, onEditRawStatus, activeQueryIndex) {
+    const activeQuery = queries[activeQueryIndex]
+    const defaultQuery = queries[0]
+    const q = activeQuery || defaultQuery
+
+    switch (view) {
+      case GRAPH:
+        return this.renderGraph(queries)
+      case TABLE:
+        return (q ? <Table query={activeQuery || defaultQuery} height={heightPixels} onEditRawStatus={onEditRawStatus} /> : <div className="generic-empty-state">Enter your query below</div>)
```

Comment:
This should be written better as a conditional. Then you wouldn't need the q const. You'd just perform the conditional logic in q as the conditional. If you're having trouble putting a conditional block in your case, make it sure there are blocks for the cases.